ISPN-14242 RESP server can send results in incorrect order if a non b…

wburns opened the pull request:
…locking client is used

https://issues.redhat.com/browse/ISPN-14242

wburns commented on server/resp/src/main/java/org/infinispan/server/resp/RespLettuceHandler.java:
```diff
@@ -19,18 +22,53 @@ public class RespLettuceHandler extends ByteToMessageDecoder {
+         if (CompletionStages.isCompletedSuccessfully(stage)) {
+            requestHandler = CompletionStages.join(stage);
+         } else {
+            log.tracef("Disabling auto read for channel %s until previous command is complete", ctx.channel());
```
This is where the backpressure for netty comes in. Basically if the command hasn't completed yet, we disable netty from reading more bytes from the socket `setAutoRead(false)` and set a variable to prevent additional commands from being processed `disabledRead = true;`. Then after the command completes we allow netty to read on the socket and for us to process commands.

We have to also finally check if there are any additional bytes read from the socket already in the netty buffer and invoke decode if so. This way if read 2 commands from the socket and the first doesn't complete right away we can process the second command without any additional activity on the socket.

wburns commented on server/resp/src/main/java/org/infinispan/server/resp/RespRequestHandler.java:
```diff
@@ -1,16 +1,109 @@
+    * @param ctx Netty context pipeline for this request
+    * @param type The command type
+    * @param arguments The remaining arguments to the command
+    * @return stage that when complete returns the new handler to instate. This stage <b>must</b> be completed on the event loop
```
The returned stage having to be completed on the event loop is very important to all of this working properly.

jabolina commented on server/resp/src/main/java/org/infinispan/server/resp/SubscriberHandler.java:
```diff
@@ -92,76 +96,78 @@ public RespRequestHandler handleRequest(ChannelHandlerContext ctx, String type,
             handler.handleRequest(ctx, type, arguments);
             break;
          case "RESET":
-            for (Iterator<Map.Entry<WrappedByteArray, PubSubListener>> iterator = specificChannelSubscribers.entrySet().iterator(); iterator.hasNext(); ) {
-               Map.Entry<WrappedByteArray, PubSubListener> entry = iterator.next();
-               PubSubListener listener = entry.getValue();
-               respServer.getCache().removeListenerAsync(listener);
-               iterator.remove();
-               sendUnsubscribe(ctx, entry.getKey().getBytes());
-            }
```
Why not unsubscribe anymore?

wburns replied:
Looking at the spec of RESET there is no reason to. https://redis.io/commands/reset/
All it says is "Aborts Pub/Sub's subscription state ([SUBSCRIBE](https://redis.io/commands/subscribe) and [PSUBSCRIBE](https://redis.io/commands/psubscribe)), when appropriate."

But I can undo this and possibly look into it in a different PR/JIRA.

wburns replied:
Looking at this more, I think I want to redo how the unsubscribe is done for the pub/sub anyways as I wasn't sure how I wanted to the error handling either.

wburns replied:
Actually looking closer, I think this is the proper thing to do as the lettuce client completely obliterates its connection.

https://github.com/lettuce-io/lettuce-core/blob/main/src/main/java/io/lettuce/core/protocol/DefaultEndpoint.java#L645

wburns replied:
I removed this as part of a different commit for [ISPN-14245](https://issues.redhat.com/browse/ISPN-14245)

wburns replied:
That is a good point about leakage, and actually there is a leak outside of this if the socket gets disconnected for example. I will add back in the removal of the listeners, however I won't call sendUnsubscribe.  I need to think of the best way to handle the client disconnect still, because that will happen with lettuce on a reset for example. I will enhance the test to verify this as well.

wburns replied:
Updated to remove all the listeners on QUIT/RESET and when the netty channel is closed. Also added a test to verify they are removed as expected (test couldn't use quit or reset commands as lettuce is doing some funky stuff there that I don't trust).

jabolina commented on server/resp/src/main/java/org/infinispan/server/resp/RespLettuceHandler.java:
```diff
@@ -19,18 +22,53 @@ public class RespLettuceHandler extends ByteToMessageDecoder {
+               log.tracef("Re-enabling auto read for channel %s as previous command is complete", ctx.channel());
+               ctx.channel().config().setAutoRead(true);
+               disabledRead = false;
+               if (t != null) {
+                  exceptionCaught(ctx, t);
+                  return;
+               }
```
By returning here we could end up losing some bytes?

wburns replied:
Ah, true. We should check the internal buffer still 👍

wburns commented:
Fixed issue with buffered commands not being fired if an exception was met in the first. Also reworked how error messages are done for pub sub multi subscribe/unsubscribe so it only sends a single error message at most.

wburns commented:
There is something left still as I enhance the concurrent test. Still looking.

wburns commented:
Fixed a bug where if multiple commands were in the netty ByteBuf before suspending reads that it would only run the one, now using `callDecode` from `ByteToMessageDecoder` so that it will repeat until it can't read anymore.

Updated the concurrent operations test a bit more resilient as it also tries a command after the retry to ensure it completes properly.

Also made sure no thread leaks were happening from the redis servers/clients not being shut down.

pruivo commented on server/resp/src/main/java/org/infinispan/server/resp/Resp3Handler.java:
```diff
@@ -53,180 +52,55 @@ public RespRequestHandler handleRequest(ChannelHandlerContext ctx, String type,
-                        }));
+            return stageToReturn(cache.getAsync(keyBytes), ctx, (innerValueBytes, t) -> {
+               if (t != null) {
+                  log.trace("Exception encountered while performing GET", t);
```
nitpick, I would probably check `log.isTraceEnabled()` in all places where `log.trace()` is used.

pruivo replied:
`really necessary` -> does it mean some intensive computation?

wburns replied:
If it requires additional allocations or computation to generate the message.

pruivo commented on server/resp/src/main/java/org/infinispan/server/resp/Resp3Handler.java:
```diff
@@ -338,16 +211,139 @@ private static CompletionStage<Long> counterIncOrDec(Cache<byte[], byte[]> cache
+      }
+      return stageToReturn(getStage.freeze(), ctx, (ignore, t) -> {
+         if (t != null) {
+            log.trace("Exception encountered while performing multiple DEL", t);
```
```suggestion
            log.trace("Exception encountered while performing multiple GET", t);
```

wburns replied:
Good catch, bad code I moved :(

pruivo commented on server/resp/src/main/java/org/infinispan/server/resp/RespLettuceHandler.java:
```diff
@@ -19,18 +22,54 @@ public class RespLettuceHandler extends ByteToMessageDecoder {
+
+               // If there is any readable bytes left before we paused make sure to try to decode, just in case
+               // if a pending message was read before we disabled auto read
+               ByteBuf buf = internalBuffer();
```
is this really required? wouldn't Netty invoke decode after re-enabling the auto read?

wburns replied:
It is required. The reason is because the decode will not be called again after reenabling auto read unless there is more socket activity.

pruivo commented on server/resp/src/main/java/org/infinispan/server/resp/RespRequestHandler.java:
```diff
@@ -1,16 +1,109 @@
-   default RespRequestHandler handleRequest(ChannelHandlerContext ctx, String type, List<byte[]> arguments) {
+public abstract class RespRequestHandler {
+   protected final CompletionStage<RespRequestHandler> myStage = CompletableFuture.completedStage(this);
+   protected final Function<Object, RespRequestHandler> myFunction = __ -> this;
```
unused

pruivo commented on server/resp/src/main/java/org/infinispan/server/resp/RespRequestHandler.java:
```diff
@@ -1,16 +1,109 @@
+            if (biConsumer != null) {
+               biConsumer.accept(result, null);
+            }
+         } catch (Throwable t) {
```
In this code path, `biConsumer.accept` is invoked twice if `biConsumer.accept(result, null);` throws an exception.
I believe you want 2 try-catch blocks:
* one around `join`
* and another around `biConsumer.accept`

wburns replied:
Yeah, the assumption was that biConsumer would not be throwing an exception. I can handle that case as you mention.

wburns replied:
Reworked this to be cleaner and more defines how the exception handling works

wburns commented on server/resp/src/main/java/org/infinispan/server/resp/Resp3AuthHandler.java:
```diff
@@ -39,51 +38,51 @@ public RespRequestHandler handleRequest(ChannelHandlerContext ctx, String type,
    private boolean handleAuthResponse(ChannelHandlerContext ctx, Subject subject) {
+      assert ctx.channel().eventLoop().inEventLoop();
       if (subject == null) {
          ctx.writeAndFlush(ctx.writeAndFlush(RespRequestHandler.stringToByteBuf("-ERR Client sent AUTH, but no password is set\r\n", ctx.alloc())));
```
No, haha. Bug from a prior PR, good catch :)

pruivo commented:
restarted CI job since it failed with OOME

pruivo commented:
merged! thanks @wburns !
